sound/soc/sof/intel/hda-pcm.c

```diff
@@ -263,6 +263,18 @@ int hda_dsp_pcm_open(struct snd_sof_dev *sdev,
 		snd_pcm_hw_constraint_mask64(substream->runtime, SNDRV_PCM_HW_PARAM_FORMAT,
 					     SNDRV_PCM_FMTBIT_S16 | SNDRV_PCM_FMTBIT_S32);
 
+	/*
+	 * Set constraint on the period time to make sure that it is larger than
+	 * the DSP buffer size. This is the maximum DMA burst size that can
+	 * happen on stream start for example.
+	 * The period size must not be smaller than this.
+	 */
+	if (spcm->stream[substream->stream].dsp_buffer_time_ms)
+		snd_pcm_hw_constraint_minmax(substream->runtime,
+			SNDRV_PCM_HW_PARAM_PERIOD_TIME,
+			spcm->stream[substream->stream].dsp_buffer_time_ms * USEC_PER_MSEC,
+			UINT_MAX);
+
```

Comment:
The constraint on BUFFER_TIME still leaves `mpv` broken, it picks 4800 (S16_LE, stereo, 48KHz = 100ms) buffer with 2 periods and it ends up in endless xrun loop.

The constraint must be on the period size, or we should make the buffer size to be at least twice of the host DMA buffer and leave the period alone?

Reply:
"When we start a playback the whole host DMA buffer is filled regardless of it's size"

Isn't this the problem statement?
I wonder why the firmware has this requirement. There's a concept of periods on the DSP side as well, so it could perfectly well start processing as soon as the first period is filled.